File: app/src/main/java/ua/hospes/rtm/ui/cars/CarsAdapter.java
package ua.hospes.rtm.ui.cars;

import android.content.res.Resources;
import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import ua.hospes.absrvadapter.AbsRecyclerAdapter;
import ua.hospes.absrvadapter.AbsRecyclerHolder;
import ua.hospes.rtm.R;
import ua.hospes.rtm.domain.cars.models.Car;
import ua.hospes.rtm.utils.UiUtils;

/**
 * @author Andrew Khloponin
 */
public class CarsAdapter extends AbsRecyclerAdapter<Car, CarsAdapter.MyHolder> {
    @Override
    public MyHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new MyHolder(parent, R.layout.item_car);
    }

    @Override
    public void onBindViewHolder(MyHolder holder, Car item, int position) {
        Resources res = holder.itemView.getResources();
        holder.number.setText(String.valueOf(item.getNumber()));
        holder.number.setTextColor(item.isBroken() ? Color.GRAY : res.getColor(item.getQuality().getColor()));
        holder.broken.setVisibility(item.isBroken() ? View.VISIBLE : View.GONE);
    }


    public class MyHolder extends AbsRecyclerHolder {
        private TextView number, broken;

        MyHolder(ViewGroup parent, int layoutId) {
            super(parent, layoutId);

            itemView.setOnClickListener(this::initClickListener);
        }


        @Override
        protected void findViews(View itemView) {
            number = UiUtils.findView(itemView, R.id.number);
            broken = UiUtils.findView(itemView, R.id.broken);
        }

        void initClickListener(View view) {
            final int position = getAdapterPosition();
            if (position == RecyclerView.NO_POSITION) return;

            if (getOnItemClickListener() == null) return;
            onItemClick(getItem(position), position);
        }
    }
}

File: app/src/main/java/ua/hospes/rtm/ui/cars/CarsAdapter.kt
package ua.hospes.rtm.ui.cars

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import ua.hospes.rtm.R
import ua.hospes.rtm.databinding.ItemCarBinding
import ua.hospes.rtm.domain.cars.Car

class CarsAdapter : ListAdapter<Car, CarsAdapter.MyHolder>(DIFF_CALLBACK) {
    var itemClickListener: ((car: Car) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyHolder =
            MyHolder(parent) { itemClickListener?.invoke(getItem(it)) }

    override fun onBindViewHolder(holder: MyHolder, position: Int) = holder.bind(getItem(position))


    inner class MyHolder(parent: ViewGroup, itemClickListener: (position: Int) -> Unit)
        : RecyclerView.ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_car, parent, false)) {

        private val binding = ItemCarBinding.bind(itemView)

        init {
            itemView.setOnClickListener {
                if (bindingAdapterPosition == RecyclerView.NO_POSITION) return@setOnClickListener
                itemClickListener.invoke(bindingAdapterPosition)
            }
        }

        fun bind(item: Car) = with(itemView) {
            binding.number.text = item.number.toString()
            when (item.broken) {
                true -> {
                    binding.number.setTextColor(Color.GRAY)
                    binding.broken.visibility = View.VISIBLE
                }
                else -> {
                    binding.number.setTextColor(resources.getColor(item.quality.color, itemView.context.theme))
                    binding.broken.visibility = View.GONE
                }
            }
        }
    }
}

private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Car>() {
    override fun areItemsTheSame(old: Car, new: Car): Boolean = old.id == new.id

    override fun areContentsTheSame(old: Car, new: Car): Boolean = old == new
}